flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/utils/MySqlSchemaUtils.java

```diff
@@ -117,6 +118,114 @@ public static List<TableId> listTables(JdbcConnection jdbc, String dbName) throw
+            }
+            return tableIds;
+        } catch (SQLException e) {
+            throw new RuntimeException("Error to list databases: " + e.getMessage(), e);
+        }
+    }
+
+    /**
+     * list all tables which contain primary key in a database.
+     *
+     * @param jdbc
+     * @param dbName
+     * @return
+     */
+    public static List<TableId> listPrimaryKeyTables(JdbcConnection jdbc, String dbName)
+            throws SQLException {
+
+        final List<TableId> tableIds = new ArrayList<>();
+
+        jdbc.query(
+                "SELECT DISTINCT TABLE_NAME FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS "
+                        + "WHERE CONSTRAINT_TYPE = 'PRIMARY KEY' AND TABLE_SCHEMA = '"
+                        + dbName
+                        + "'",
+                rs -> {
+                    while (rs.next()) {
+                        tableIds.add(TableId.tableId(dbName, rs.getString(1)));
+                    }
+                });
+
+        LOG.info("\t list of available tables are: {}", tableIds);
+
+        return tableIds;
+    }
+
+    /**
+     * list all tables without primary key in a database.
+     *
+     * @param sourceConfig
+     * @param dbName
+     * @return
+     */
+    public static List<TableId> listNoPrimaryKeyTables(
+            MySqlSourceConfig sourceConfig, String dbName) {
+        try (MySqlConnection jdbc = createMySqlConnection(sourceConfig)) {
+            List<String> databases =
+                    dbName != null ? Collections.singletonList(dbName) : listDatabases(jdbc);
+
+            List<TableId> tableIds = new ArrayList<>();
+            for (String database : databases) {
+                tableIds.addAll(listNoPrimaryKeyTables(jdbc, database));
+            }
+            return tableIds;
+        } catch (SQLException e) {
+            throw new RuntimeException("Error to list databases: " + e.getMessage(), e);
+        }
+    }
+
+    /**
+     * list all tables without primary key in a database.
+     *
+     * @param jdbc
+     * @param dbName
+     * @return
+     */
+    public static List<TableId> listNoPrimaryKeyTables(JdbcConnection jdbc, String dbName)
+            throws SQLException {
+        final List<TableId> tableIds = new ArrayList<>();
+
+        jdbc.query(
+                "select tab.table_name\n"
+                        + " from information_schema.tables tab\n"
+                        + " left join information_schema.table_constraints tco\n"
+                        + "          on tab.table_schema = tco.table_schema\n"
+                        + "          and tab.table_name = tco.table_name\n"
+                        + "          and tco.constraint_type = 'PRIMARY KEY'\n"
+                        + " where tco.constraint_type is null\n"
+                        + "      and tab.table_type = 'BASE TABLE'\n"
+                        + "      and tab.table_schema = '"
+                        + dbName
```

Comment:
Hi @dufeng1010 , I wonder why this SQL is quiet different with the listPrimaryKeyTables method.

Reply:
this method is get all tables without primary key. 
listPrimarykeyTables method is get all tables with primary key

Reply:
Get it. It's better to unify the case.